Migrate custom-xcom-backends guide (#1127)
astronomer/docs: guides/custom-xcom-backends.md

---

> title: "Custom XCom Backends"

Reviewer: ```suggestion
title: "Custom XCom backends"
```

> sidebar_label: "Custom XCom Backends"

Reviewer: ```suggestion
sidebar_label: "Custom XCom backends"
```

description: "Creating a custom XCom backend with Airflow 2.0."
id: custom-xcom-backends
---

One of the most significant features in the [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) release was the [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html#taskflow-api). The TaskFlow API was introduced to solve the challenge of explicitly passing messages between Airflow tasks. The TaskFlow API abstracts the task and dependency management layer and improves the experience of working with XComs.

> One of the features of the TaskFlow API that increases the flexibility of XComs is support for a [custom XCom backend](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=xcom#custom-xcom-backend). This means that rather than store XComs in the Airflow metadata database by default, you can push and pull XComs to and from an external system such as Amazon S3, Google Cloud Storage (GCS), or Hadoop Distributed File System (HDFS). You can also implement your own serialization and deserialization methods to define how XComs are handled. 

Reviewer: ```suggestion
As part of this update, support was introduced for configuring [custom XCom backends](https://airflow.apache.org/docs/apache-airflow/stable/concepts.html?highlight=xcom#custom-xcom-backend). This means that rather than store XComs in the Airflow metadata database by default, you can push and pull XComs to and from an external system such as Amazon S3, Google Cloud Storage (GCS), or Hadoop Distributed File System (HDFS). You can also implement your own serialization and deserialization methods to define how XComs are handled. 
```

            hook    = S3Hook()
            key     = result.replace(S3XComBackend.PREFIX, "")
            filename = hook.download_file(
                key=key,
                bucket_name=S3XComBackend.BUCKET_NAME,
                local_path="/tmp"
            )
            result = pd.read_csv(filename)
        return result

```

To add this class to your Airflow environment, complete the following steps:

1. Create a folder and a file in your host directory and paste the `S3XComBackend` class code above into the file:

    ```docker
    # Create a folder/file in your host directory
    include/s3_xcom_backend.py
    ```

2. Mount the folder you created in step 1 as a Docker volume in your Airflow project. If you're using docker-compose, you can add the following to your `docker-compose.yaml` file:

    ```docker
    volumes:
    - ./include:/opt/airflow/include
    ```

3. Specify your XCom backend by defining the `AIRFLOW__CORE__XCOM_BACKEND` Airflow environment variable. If you're using docker-compose, you can add the following to your `docker-compose.yaml` file:

    ```docker
    environment:
    - AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend
    ```

> To run this docker-compose example, you need to set your Python Path in your `docker-compose.yaml` file. For example, `PYTHONPATH: /opt/airflow/`.

Reviewer: ```suggestion
    To run this docker-compose example, you need to set your Python Path in your `docker-compose.yaml` file. For example, `PYTHONPATH: /opt/airflow/`.
```

When you restart Airflow and run a DAG with pandas dataframe XComs, the XComs get pushed to your Amazon S3 backend as shown in the following image:

![XCom Backend S3 XComs](/img/guides/xcom_backend_s3_xcoms.png)

If you've changed your XCom Backend class by modifying `s3_xcom_backend.py`, you'll need to restart Airflow with `docker-compose down && docker-compose up -d` to implement the changes.

## Using XCom backends on Astro

In this example, you'll use the Great Expectations use case for storing operator results XComs. If you haven't installed the Astronomer CLI, install it using the [CLI Quickstart](https://docs.astronomer.io/astro/install-cli).

1. Configure your backend. See [Set up an XCom backend](#set-up-an-xcom-backend).
2. Open an existing Astronomer project, or initialize a new one using `astro dev init`. 
3. Add the XCom backend Python file with serialization and deserialization methods to the `include/` directory of your Astronomer project.

> 4. Add the `AIRFLOW__CORE_XCOM_BACKEND` environment variable to your Astronomer project. For this example, it be `AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend`. There are a few ways to add an environment variable to your Astronomer Deployment, which are detailed in the [Environment Variables guide](https://www.astronomer.io/docs/cloud/stable/deploy/environment-variables). In this example, you added the variable to your Dockerfile.

Reviewer: ```suggestion
4. Add the `AIRFLOW__CORE_XCOM_BACKEND` environment variable to your Astronomer project. For this example, it be `AIRFLOW__CORE__XCOM_BACKEND=include.s3_xcom_backend.S3XComBackend`. There are a few ways to add an environment variable to your Astronomer Deployment, which are detailed in the [Environment Variables guide](https://docs.astronomer.io/astro/environment-variables). This example assumes that you added the variable to your Dockerfile.
```
